src/client/content_mapblock.cpp

```diff
@@ -1039,8 +1075,23 @@ void MapblockMeshGenerator::drawPlantlikeNode()
 
 void MapblockMeshGenerator::drawPlantlikeRootedNode()
 {
+	u8 facedir = 0;
+
+	if (f->param_type_2 == CPT2_FACEDIR ||
+			f->param_type_2 == CPT2_COLORED_FACEDIR) {
+		facedir = n.getFaceDir(nodedef);
+	} else if (f->param_type_2 == CPT2_WALLMOUNTED ||
+			f->param_type_2 == CPT2_COLORED_WALLMOUNTED) {
+		// Convert wallmounted to 6dfacedir.
+		// When cache enabled, it is already converted.
+		facedir = n.getWallMounted(nodedef);
+		if (!enable_mesh_cache)
+			facedir = wallmounted_to_facedir[facedir];
+	}
+	
 	useTile(0, MATERIAL_FLAG_CRACK_OVERLAY, 0, true);
-	origin += v3f(0.0, BS, 0.0);
+	v3s16 dir = g_6dirs[facedir_to_dir[facedir%32]];
```

Comment:
Similarily this doesn't need a new table with 24 entries:
```cpp
	const static u8 face_axis[] = {1, 0, 3, 2, 5, 4};
	v3s16 dir = g_6dirs[face_axis[facedir >> 2]];
	origin += v3f(dir.X * BS, dir.Y * BS, dir.Z * BS);
	p += dir;
	if (data->m_smooth_lighting) {
		getSmoothLightFrame();
	} else {
		MapNode ntop = data->m_vmanip.getNodeNoEx(blockpos_nodes + p);
		light = LightPair(getInteriorLight(ntop, 1, nodedef));
	}
	drawPlantlike();
	p -= dir;
```
(also changed: adjust `p` according to dir)

Reply:
Fixed.
I have deliver this from [lua facedir_to_dir function](https://github.com/minetest/minetest/blob/90a7bd6a0afb6509e96bcb373f95b448ee7f3b1d/builtin/game/item.lua#L103)

```lua
-- Mapping from facedir value to index in facedir_to_dir.
        local facedir_to_dir_map = {
        	[0]=1, 2, 3, 4,
        	5, 2, 6, 4,
        	6, 2, 5, 4,
        	1, 5, 3, 6,
        	1, 6, 3, 5,
        	1, 4, 3, 2,
        }
```
And it does not work well, so I changed it and miss the optimization possibility.